rust/gui-client/src-tauri/src/client/updates.rs

```diff
@@ -1,18 +1,199 @@
+            }
+            CheckerStep::CheckNetwork => {
+                let release = match check().await {
+                    Ok(x) => x,
+                    Err(error) => {
+                        tracing::error!(?error, "Couldn't check website for update");
+                        fsm.handle_failed_check();
+                        continue;
+                    }
+                };
+                if let Some(release) = fsm.handle_check(release) {
+                    ctlr_tx
+                        .send(ControllerRequest::UpdateAvailable(release.clone()))
+                        .await?;
+                    write_latest_release_file(release).await?;
+                }
+            }
+            CheckerStep::WaitInterval => {
+                interval.tick().await;
+                fsm.handle_wake();
+            }
+        }
+    }
+}
+
+async fn read_latest_release_file() -> Option<Release> {
+    tokio::fs::read_to_string(version_file_path().ok()?)
+        .await
+        .ok()
+        .as_deref()
+        .map(serde_json::from_str)
+        .transpose()
+        .ok()
+        .flatten()
+}
+
+async fn write_latest_release_file(release: Release) -> Result<()> {
+    // `atomicwrites` is sync so use `spawn_blocking` so we don't block an
+    // executor thread
+    tokio::task::spawn_blocking(move || {
+        let f = atomicwrites::AtomicFile::new(
+            version_file_path()?,
+            atomicwrites::OverwriteBehavior::AllowOverwrite,
+        );
+        f.write(|f| f.write_all(serde_json::to_string(&release)?.as_bytes()))?;
+        Ok::<_, anyhow::Error>(())
+    })
+    .await??;
+    Ok(())
+}
+
+struct Checker {
+    latest_seen: Option<Version>,
+    ours: Version,
+    state: CheckerStep,
+}
+
+/// Output events for I/O and internal state of the update checker state machine.
+///
+/// The machine is simple enough that they happen to be identical:
+///
+/// - CheckFile
+/// - WaitRandom
+/// - CheckNetwork
+/// - WaitInterval
+/// - (Loop back to CheckNetwork)
+#[derive(Clone, Copy)]
+enum CheckerStep {
+    /// Check the disk to see what was the latest version we remember seeing.
+    /// This allows us to show the notification quicker without hitting the network right at startup.
+    CheckFile,
+    /// Wait a random amount of time to break up thundering herds.
+    WaitRandom,
+    /// Check the latest version from the Firezone website and write it to disk.
+    CheckNetwork,
+    /// Wait approximately a day using `tokio::time::interval`.
+    WaitInterval,
+}
+
+impl Checker {
+    fn new(ours: Version) -> Self {
+        Self {
+            latest_seen: None,
+            ours,
+            state: CheckerStep::CheckFile,
+        }
+    }
+
+    /// Call this when we just checked the network or the file
+    ///
+    /// Returns the release if we should alert the UI about this release
+    #[must_use]
+    fn handle_check(&mut self, release: Release) -> Option<Release> {
```

Comment:
In connlib, most of our state machines don't have an immediate return value but instead, they emit additional events. This usually makes the "driver" of the SM easier to write because you need to handle possible events / return values only in a single place.

This matters especially once you start having more functions that feed new data into the SM.